pkg/metadata/pod_hosts.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/prometheus/common/model"
+)
+
+var (
+	initLocalHostname sync.Once
+	hostname          string
+)
+
+// PodHosts provide host ip in default
+// and will provide pod_ip and pod if pid is a pod.
+func PodHosts() Provider {
+	return &StatelessProvider{"hosts", func(pid int) (model.LabelSet, error) {
+		initLocalHostname.Do(func() {
+			local, err := getExternalIPAndHost(1)
+			if err != nil {
+				// this should not happen
+				panic(err)
+			}
+			hostname = local.hostname
+		})
+		e, err := getExternalIPAndHost(pid)
+		if err != nil {
+			return nil, err
+		}
+		if hostname != e.hostname {
+			// pod
+			return model.LabelSet{
+				"pod_ip": model.LabelValue(e.ip),
+				"pod":    model.LabelValue(e.hostname),
+			}, nil
+		}
+		return nil, nil
+	}}
+}
+
+type hostEntry struct {
+	ip       string
+	hostname string
+}
+
+func getExternalIPAndHost(pid int) (*hostEntry, error) {
+	data, err := os.ReadFile(fmt.Sprintf("/proc/%d/root/etc/hosts", pid))
+	if err != nil {
+		return nil, err
+	}
+	hostEntrys, err := parseHosts(bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	if len(hostEntrys) == 0 {
+		return nil, fmt.Errorf("read hosts file result nil, raw:%s", string(data))
+	}
+	return &hostEntrys[len(hostEntrys)-1], nil
+}
+
+func parseHosts(r io.Reader) ([]hostEntry, error) {
```

Comment:
```suggestion
func parseHosts(r io.Reader, entries []hostEntry)  error {
```

This function can take the resulting slice from the caller to avoid additional escape to the heap.

Reply:
Do you mean prealloc entries?
Like parseHosts(r, make([]entries, 0, 3))?

But it is not easy to know the entries cap.

Reply:
Never mind, it's a micro-optimization.
But this is what I've meant 👉 https://blog.devgenius.io/in-depth-analysis-of-golang-memory-escape-edfbfb856913